File: multi-instance-deploy/templates/matrix-bot/set_display_name.py
#!/usr/bin/env python3
"""
Set Matrix Bot Display Name - Globally or Per-Room
====================================================
Usage:
    python3 set_display_name.py "My Cool Name"              # Set global display name
    python3 set_display_name.py "Room Bot" --room ROOM_ID   # Set per-room display name
    python3 set_display_name.py --reset                     # Reset to default (user ID)

Author: Agent Zero
Version: 1.0
"""

import asyncio
import argparse
import os
import sys
from pathlib import Path
from dotenv import load_dotenv
from nio import AsyncClient

# Load environment
BOT_DIR = Path(__file__).parent
load_dotenv(BOT_DIR / ".env")

HOMESERVER_URL = os.getenv("MATRIX_HOMESERVER_URL", "http://localhost:8008")
USER_ID = os.getenv("MATRIX_USER_ID", "")
ACCESS_TOKEN = os.getenv("MATRIX_ACCESS_TOKEN", "")
DEVICE_ID = os.getenv("MATRIX_DEVICE_ID", "AgentZeroBot")


async def set_global_display_name(name: str):
    """Set display name globally (all rooms)."""
    if not all([HOMESERVER_URL, USER_ID, ACCESS_TOKEN]):
        print("❌ Error: Missing required environment variables")
        print("   Need: MATRIX_HOMESERVER_URL, MATRIX_USER_ID, MATRIX_ACCESS_TOKEN")
        sys.exit(1)
    
    client = AsyncClient(
        homeserver=HOMESERVER_URL,
        user=USER_ID,
        device_id=DEVICE_ID,
    )
    client.access_token = ACCESS_TOKEN
    client.user_id = USER_ID
    
    try:
        response = await client.set_displayname(name)
        if hasattr(response, 'displayname'):
            print(f"✅ Global display name set to: {name}")
        else:
            print(f"⚠️  Response: {response}")
    except Exception as e:
        print(f"❌ Error setting display name: {e}")
    finally:
        await client.close()


async def set_room_display_name(room_id: str, name: str):
    """Set display name for a specific room."""
    if not all([HOMESERVER_URL, USER_ID, ACCESS_TOKEN]):
        print("❌ Error: Missing required environment variables")
        sys.exit(1)
    
    client = AsyncClient(
        homeserver=HOMESERVER_URL,
        user=USER_ID,
        device_id=DEVICE_ID,
    )
    client.access_token = ACCESS_TOKEN
    client.user_id = USER_ID
    
    try:
        content = {
            "displayname": name,
            "membership": "join"
        }
        response = await client.room_put_state(
            room_id=room_id,
            event_type="m.room.member",
            content=content,
            state_key=USER_ID
        )
        if hasattr(response, 'event_id'):
            print(f"✅ Room display name set to: {name}")
            print(f"   Room: {room_id}")
        else:
            print(f"⚠️  Response: {response}")
    except Exception as e:
        print(f"❌ Error setting room display name: {e}")
    finally:
        await client.close()


async def reset_display_name():
    """Reset display name to default (user ID localpart)."""
    default_name = USER_ID.split(":")[0].lstrip("@") if USER_ID else "Agent"
    await set_global_display_name(default_name)


async def main():
    parser = argparse.ArgumentParser(
        description="Set Matrix bot display name globally or per-room"
    )
    parser.add_argument(
        "name",
        nargs="?",
        help="Display name to set"
    )
    parser.add_argument(
        "--room", "-r",
        help="Room ID to set per-room display name"
    )
    parser.add_argument(
        "--reset",
        action="store_true",
        help="Reset to default display name"
    )
    parser.add_argument(
        "--list",
        action="store_true",
        help="Show current configuration"
    )
    
    args = parser.parse_args()
    
    if args.list:
        print("📋 Current Configuration:")
        print(f"   Homeserver: {HOMESERVER_URL}")
        print(f"   User ID: {USER_ID}")
        print(f"   Device ID: {DEVICE_ID}")
        return
    
    if args.reset:
        await reset_display_name()
        return
    
    if not args.name:
        print("❌ Error: Please provide a display name or use --reset")
        parser.print_help()
        sys.exit(1)
    
    if args.room:
        await set_room_display_name(args.room, args.name)
    else:
        await set_global_display_name(args.name)


if __name__ == "__main__":
    asyncio.run(main())

File: multi-instance-deploy/templates/matrix-bot/rust/src/bin/set_display_name.rs
use anyhow::{anyhow, Context, Result};
use dotenvy::dotenv;
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde_json::json;

#[derive(Clone, Debug)]
struct Config {
    homeserver_url: String,
    user_id: String,
    access_token: String,
    device_id: String,
}

impl Config {
    fn from_env() -> Self {
        Self {
            homeserver_url: std::env::var("MATRIX_HOMESERVER_URL")
                .unwrap_or_else(|_| "http://localhost:8008".to_string()),
            user_id: std::env::var("MATRIX_USER_ID").unwrap_or_default(),
            access_token: std::env::var("MATRIX_ACCESS_TOKEN").unwrap_or_default(),
            device_id: std::env::var("MATRIX_DEVICE_ID").unwrap_or_else(|_| "AgentZeroBot".to_string()),
        }
    }
}

fn usage() {
    eprintln!("Usage:");
    eprintln!("  set-display-name-rust \"My Name\"");
    eprintln!("  set-display-name-rust \"Room Bot\" --room ROOM_ID");
    eprintln!("  set-display-name-rust --reset");
    eprintln!("  set-display-name-rust --list");
}

#[tokio::main]
async fn main() {
    let _ = dotenv();
    if let Err(e) = run().await {
        eprintln!("Error: {e}");
        std::process::exit(1);
    }
}

async fn run() -> Result<()> {
    let cfg = Config::from_env();
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() {
        usage();
        return Err(anyhow!("missing arguments"));
    }

    if args.iter().any(|a| a == "--list") {
        println!("Current Configuration:");
        println!("  Homeserver: {}", cfg.homeserver_url);
        println!("  User ID: {}", cfg.user_id);
        println!("  Device ID: {}", cfg.device_id);
        return Ok(());
    }

    if cfg.homeserver_url.is_empty() || cfg.user_id.is_empty() || cfg.access_token.is_empty() {
        return Err(anyhow!(
            "Missing required env vars: MATRIX_HOMESERVER_URL, MATRIX_USER_ID, MATRIX_ACCESS_TOKEN"
        ));
    }

    let mut name: Option<String> = None;
    let mut room_id: Option<String> = None;
    let mut reset = false;

    let mut i = 0usize;
    while i < args.len() {
        match args[i].as_str() {
            "--room" | "-r" => {
                if i + 1 >= args.len() {
                    return Err(anyhow!("--room requires ROOM_ID"));
                }
                room_id = Some(args[i + 1].clone());
                i += 2;
            }
            "--reset" => {
                reset = true;
                i += 1;
            }
            x if x.starts_with("--") => {
                return Err(anyhow!("unknown flag: {}", x));
            }
            value => {
                if name.is_none() {
                    name = Some(value.to_string());
                    i += 1;
                } else {
                    return Err(anyhow!("unexpected argument: {}", value));
                }
            }
        }
    }

    if reset {
        let default_name = cfg
            .user_id
            .split(':')
            .next()
            .unwrap_or("Agent")
            .trim_start_matches('@')
            .to_string();
        set_global_display_name(&cfg, &default_name).await?;
        println!("Global display name reset to: {}", default_name);
        return Ok(());
    }

    let name = name.ok_or_else(|| anyhow!("Please provide a display name or use --reset"))?;
    if let Some(room) = room_id {
        set_room_display_name(&cfg, &room, &name).await?;
        println!("Room display name set to: {}", name);
        println!("Room: {}", room);
    } else {
        set_global_display_name(&cfg, &name).await?;
        println!("Global display name set to: {}", name);
    }
    Ok(())
}

fn client(cfg: &Config) -> Result<reqwest::Client> {
    let mut headers = HeaderMap::new();
    headers.insert(
        AUTHORIZATION,
        HeaderValue::from_str(&format!("Bearer {}", cfg.access_token)).context("invalid MATRIX_ACCESS_TOKEN")?,
    );
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    reqwest::Client::builder()
        .default_headers(headers)
        .build()
        .context("build reqwest client")
}

async fn set_global_display_name(cfg: &Config, name: &str) -> Result<()> {
    let user_enc = urlencoding::encode(&cfg.user_id);
    let url = format!(
        "{}/_matrix/client/v3/profile/{}/displayname",
        cfg.homeserver_url.trim_end_matches('/'),
        user_enc
    );
    let resp = client(cfg)?.put(url).json(&json!({ "displayname": name })).send().await?;
    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(anyhow!("set display name failed {status}: {body}"));
    }
    Ok(())
}

async fn set_room_display_name(cfg: &Config, room_id: &str, name: &str) -> Result<()> {
    let room_enc = urlencoding::encode(room_id);
    let user_enc = urlencoding::encode(&cfg.user_id);
    let url = format!(
        "{}/_matrix/client/v3/rooms/{}/state/m.room.member/{}",
        cfg.homeserver_url.trim_end_matches('/'),
        room_enc,
        user_enc
    );
    let payload = json!({
        "displayname": name,
        "membership": "join"
    });
    let resp = client(cfg)?.put(url).json(&payload).send().await?;
    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(anyhow!("set room display name failed {status}: {body}"));
    }
    Ok(())
}
